File: LeetCode/0301.Remove-Invalid-Parentheses/Remove-Invalid-Parentheses.cpp
class Solution {
public:
    vector<string> removeInvalidParentheses(string s) {
        vector<string> res;
        queue<string> q{{s}};
        unordered_set<string> seen{{s}};
        bool found = false;
        while (!q.empty()) {
            string cur = q.front(); q.pop();
            if (isValid(cur)) {
                res.push_back(cur);
                found = true;
            }
            if (found) continue;
            for (int i = 0; i < cur.size(); i++) {
                if (cur[i] == '(' || cur[i] == ')') {
                    string next = cur.substr(0, i) + cur.substr(i + 1);
                    if (!seen.count(next)) {
                        q.push(next);
                        seen.insert(next);
                    }
                }
            }
        }
        return res;
    }
    bool isValid(string s) {
        int cnt = 0;
        for (int i = 0; i < s.size(); i++) {
            if (s[i] == '(') cnt++;
            if (s[i] == ')') {
                if (cnt == 0) return false;
                cnt--;
            }
        }
        return cnt == 0;
    }
};

File: LeetCode/0301.Remove-Invalid-Parentheses/Remove-Invalid-Parentheses.py
class Solution(object):
    def removeInvalidParentheses(self, s):
        """
        :type s: str
        :rtype: List[str]
        """
        def is_valid(s):
            cnt = 0
            for c in s:
                if c == "(": cnt += 1
                if c == ")":
                    if cnt <= 0: return False
                    cnt -= 1
            return cnt == 0
        
        def bfs(s):
            visited = set([s])
            queue = collections.deque([s])
            found = False
            ans = []
            while queue:
                cur = queue.popleft()
                #print cur
                if is_valid(cur):
                    ans.append(cur)
                    found = True
                if found == True:
                    continue
                for i in range(len(cur)):
                    if cur[i] != "(" and cur[i] != ")":
                        continue
                    new = cur[:i] + cur[i + 1:]
                    #print new, visited
                    if new not in visited:
                        queue.append(new)
                        visited.add(new)
            return ans
        
        def dfs(s, start, left, right, ans):
            if left == 0 and right == 0 and is_valid(s):
                ans.append(s)
                return
            for i in range(start, len(s)):
                if i > start and s[i] == s[i - 1]: continue
                if left > 0 and s[i] == "(":
                    dfs(s[:i] + s[i + 1:], i, left - 1, right, ans)
                if right > 0 and s[i] == ")":
                    dfs(s[:i] + s[i + 1:], i, left, right - 1, ans)
        #return bfs(s)
        left = right = 0
        for c in s:
            if c == "(": left += 1
            if c == ")":
                if left: left -= 1
                else: right += 1
        ans = []
        dfs(s, 0, left, right, ans)
        return ans
    
        #return bfs(s)
